### ISSUE TITLE
[Performance] The CUDA Stream cannot be set through Python API

### ISSUE THREAD

**Issue by gedoensmax** [microsoft/onnxruntime#19094]
### Describe the issue

I want to set a CUDA Stream that I can acquire in python through torch or cuda for python. Ideally I want to asynchronously submit work to that stream to pipeline work better.

With CUDA EP I am able to do this by setting:
```
        run_options = ort.RunOptions()
        run_options.add_run_config_entry("disable_synchronize_execution_providers", "1")
```
On TRT EP this has no effect due to checking if an external stream is being used as well: https://github.com/microsoft/onnxruntime/blob/5f3113ecd612b4ef070d5a3c8a586d725cfdb13b/onnxruntime/core/providers/tensorrt/tensorrt_execution_provider.cc#L1760

It would be great to set a user stream through the Python API ! 

### To reproduce

Following code fails and I am not aware of any alternative API. If there is any it is not documented I believe.
```
ort_sess = ort.InferenceSession('model.onnx', providers=('CUDAExecutionProvider', {
                 'user_compute_stream': cuda_stream   
})
```

### Platform

Linux

### OS Version

Ubuntu

### ONNX Runtime Installation

Built from Source

### ONNX Runtime Version or Commit ID

main

### ONNX Runtime API

Python

### Architecture

X64

### Execution Provider

CUDA, TensorRT

### Is this a quantized model?

No

**Comment by gedoensmax** [comment]
@chilo-ms Do you know the Python API architecture a bit better ? It should be a quick thing to add, but I don't really  know the details.

**Comment by hariharans29** [comment]
Currently, it seems like the provider options in Python are limited to just string types (https://github.com/microsoft/onnxruntime/blob/9da3e36138dd24377fbb0b4022d891b3baf07b84/onnxruntime/python/onnxruntime_pybind_state.cc#L1149C30-L1149C44). It does need enhancing to support other provider option types. 

On a different note, it seems like the CUDA EP and TRT EP should be consistent on how the above run option is dealt with (which I think is the core issue here)

**Comment by gedoensmax** [comment]
The core issue is that not being able to set the stream makes it complicated to use this asynchronous with e.g. PyTorch on the site. You would want to set the compute stream to get rid of unneeded synchronizations. 
But yes I agree a consistent behavior would be great.

**Comment by tianleiwu** [comment]
I think it is easy to implement it. Only need add some a few lines of code to parse user_compute_stream from string to a pointer, similar to `gpu_external_alloc`:
https://github.com/microsoft/onnxruntime/blob/f3402de01e732283283aaa208022d6c7ae85ca4a/onnxruntime/core/providers/cuda/cuda_execution_provider_info.cc#L68-L76
I could add a draft implementation later.

**Comment by chilo-ms** [comment]
@gedoensmax
 Current TRT EP only considers "run_options.add_run_config_entry("disable_synchronize_execution_providers", "1")" for user provided CUDA stream case.
 For ORT managed CUDA stream case, we don't support the configuration for disabling the sync and, instead, TRT EP calls sync stream api after every enqueueV3(). The reason is if we don't do it, there is a concurrent issue is multiple threads are running the InferenceSession::Run() where multiple streams are created internally and the sync across different stream is needed.
 see here https://github.com/microsoft/onnxruntime/pull/19301

**Comment by gedoensmax** [comment]
Yeah I get that, but I would CUDA EP to behave the same in that case.

**Comment by chilo-ms** [comment]
Yeah, then we need to discuss this internally
